File: server/src/main/java/com/toddway/firebaserx/FirebaseRx.java
package com.toddway.firebaserx;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Query;
import com.google.firebase.tasks.Task;

import java.util.Map;

import rx.Observable;

/**
 * Created by tway on 2/7/17.
 */

public class FirebaseRx {

    public static <T> Observable<T> observeTask(Task<T> task) {
        return FirebaseRxKt.observeTask(task);
    }

    public static Observable<Void> observeAddChild(DatabaseReference ref, Object object) {
        return FirebaseRxKt.observeAddChild(ref, object);
    }

    public static Observable<DataSnapshot> observeValue(Query query) {
        return FirebaseRxCommonKt.observeValue(query);
    }

    public static <T> Observable<T> observeValue(Query query, Class<T> type) {
        return FirebaseRxCommonKt.observeValue(query, type);
    }

    public static <T> Observable<Map<String, T>> observeChildMap(Query query, Class<T> type) {
        return FirebaseRxCommonKt.observeChildMap(query, type);
    }

    public static <T> Observable<Map<String, T>> observeChildMap(Query query, Class<T> type, Iterable<String> keys) {
        return FirebaseRxCommonKt.observeChildMap(query, type, keys);
    }
}

File: server/src/main/java/com/toddway/firebaserx/FirebaseRx.kt
package com.toddway.firebaserx

import com.google.firebase.database.DatabaseReference
import com.google.firebase.tasks.Task
import rx.Observable
import rx.Subscriber

/**
 * Created by tway on 8/4/17.
 */

fun <T> Task<T>.observeTask(): Observable<T> {
    return Observable.create<T> { subscriber ->
        val listener = { task : Task<T> -> taskComplete(subscriber, task) }
        this.addOnCompleteListener(listener)
    }
}

fun DatabaseReference.observeAddChild(value: Any): Observable<Void> {
    return Observable.create<Void> { subscriber ->
        val key = push().key
        child(key).setValue(value).addOnCompleteListener { task -> taskComplete(subscriber, task) }
    }
}

fun DatabaseReference.observeCopyValue(targetRef: DatabaseReference): Observable<Void> {
    return observeValue().take(1)
            .flatMap {
                if (it.exists()) targetRef.setValue(it.value).observeTask()
                else Observable.just(null)
            }
}

fun DatabaseReference.observeMoveValue(targetRef: DatabaseReference): Observable<Void> {
    return observeCopyValue(targetRef)
            .doOnNext { removeValue().observeTask().subscribe() }
}

private fun <T> taskComplete(subscriber: Subscriber<in T>, task: Task<T>) {
    if (!subscriber.isUnsubscribed) {
        if (task.isSuccessful) {
            subscriber.onNext(task.result)
            subscriber.onCompleted()
        } else {
            subscriber.onError(task.exception)
        }
    }
}